File: vs_semester-5/oop_lab-1/task-1/Laptop.cpp
#include "Laptop.h"

Laptop::Laptop(
  std::wstring mark,
  std::wstring model,
  float CPU_clockRate,
  float GPU_clockRate,
  int RAM_size,
  int GPU_RAM_size
) {
  Laptop::mark = mark;
  Laptop::model = model;
  Laptop::CPU_clockRate = CPU_clockRate;
  Laptop::GPU_clockRate = GPU_clockRate;
  Laptop::RAM_size = RAM_size;
  Laptop::GPU_RAM_size = GPU_RAM_size;
}

Laptop::Laptop(const Laptop& obj) {}

Laptop::~Laptop() {}

std::wstring Laptop::getMark() {
  return mark;
}

std::wstring Laptop::getModel() {
  return model;
}

float Laptop::getCPUcr() {
  return CPU_clockRate;
}

float Laptop::getGPUcr() {
  return GPU_clockRate;
}

int Laptop::getRAMs() {
  return RAM_size;
}

int Laptop::getGPUrs() {
  return GPU_RAM_size;
}

std::wstring Laptop::setMark(std::wstring value) {
  mark = value;
  return value;
}

std::wstring Laptop::setModel(std::wstring value) {
  model = value;
  return value;
}

float Laptop::setCPUcr(float value) {
  CPU_clockRate = value;
  return value;
}

float Laptop::setGPUcr(float value) {
  GPU_clockRate = value;
  return value;
}

int Laptop::setRAMs(int value) {
  RAM_size = value;
  return value;
}

int Laptop::setGPUrs(int value) {
  GPU_RAM_size = value;
  return value;
}

Laptop& Laptop::operator=(const Laptop& ob) {
  this->mark = ob.mark;
  this->model = ob.model;
  this->CPU_clockRate = ob.CPU_clockRate;
  this->GPU_clockRate = ob.GPU_clockRate;
  this->RAM_size = ob.RAM_size;
  this->GPU_RAM_size = ob.GPU_RAM_size;

  return *this;
}

Laptop::Laptop() {
  Laptop::mark = L"Empty";
  Laptop::model = L"Empty";
  Laptop::CPU_clockRate = 0;
  Laptop::GPU_clockRate = 0;
  Laptop::RAM_size = 0;
  Laptop::GPU_RAM_size = 0;
}

bool operator==(const Laptop& ob_l, const Laptop& ob_r) {
  bool b =
    (ob_l.mark == ob_r.mark) &&
    (ob_l.model == ob_r.model) &&
    (ob_l.CPU_clockRate == ob_r.CPU_clockRate) &&
    (ob_l.GPU_clockRate == ob_r.GPU_clockRate) &&
    (ob_l.RAM_size == ob_r.RAM_size) &&
    (ob_l.GPU_RAM_size == ob_r.GPU_RAM_size);
  return b;
}

bool operator!=(const Laptop& ob_l, const Laptop& ob_r) {
  bool b =
    (ob_l.mark != ob_r.mark) &&
    (ob_l.model != ob_r.model) &&
    (ob_l.CPU_clockRate != ob_r.CPU_clockRate) &&
    (ob_l.GPU_clockRate != ob_r.GPU_clockRate) &&
    (ob_l.RAM_size != ob_r.RAM_size) &&
    (ob_l.GPU_RAM_size != ob_r.GPU_RAM_size);
  return b;
}

bool operator>(const Laptop& ob_l, const Laptop& ob_r) {
  bool b;
  if (ob_l.mark > ob_r.mark) b = true;
  else if (ob_l.model > ob_r.model)  b = true;
  else if (ob_l.CPU_clockRate > ob_r.CPU_clockRate)  b = true;
  else if (ob_l.GPU_clockRate > ob_r.GPU_clockRate)  b = true;
  else if (ob_l.RAM_size > ob_r.RAM_size)  b = true;
  else if (ob_l.GPU_RAM_size > ob_r.GPU_RAM_size)  b = true;
  else b = false;

  return b;
}

bool operator<(const Laptop& ob_l, const Laptop& ob_r) {
  bool b;
  if (ob_l.mark < ob_r.mark) b = true;
  else if (ob_l.model < ob_r.model)  b = true;
  else if (ob_l.CPU_clockRate < ob_r.CPU_clockRate)  b = true;
  else if (ob_l.GPU_clockRate < ob_r.GPU_clockRate)  b = true;
  else if (ob_l.RAM_size < ob_r.RAM_size)  b = true;
  else if (ob_l.GPU_RAM_size < ob_r.GPU_RAM_size)  b = true;
  else b = false;

  return b;
}

std::wostream& operator<<(std::wostream& out, const Laptop& ob) {
  out << ob.model << L"\n\r";
  out << ob.mark << L"\n\r";
  out << ob.CPU_clockRate << L"\n\r";
  out << ob.GPU_clockRate << L"\n\r";
  out << ob.RAM_size << L"\n\r";
  out << ob.GPU_RAM_size << L"\n\r";
  out << L"\n\r";

  return out;
}

std::wistream& operator>>(std::wistream& in, Laptop& ob) {
  in >> ob.model;
  in >> ob.mark;
  in >> ob.CPU_clockRate;
  in >> ob.GPU_clockRate;
  in >> ob.RAM_size;
  in >> ob.GPU_RAM_size;

  return in;
}

File: semester-5_vs/oop_lab-4/task-1_cs/Laptop.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace task_1_cs {
  public interface IComparable<T> {
    int CompareTo(T obj);
  }

  class Laptop : IComparable<Laptop> {
    int IComparable<Laptop>.CompareTo(Laptop obj) {
      // 1 поле
      if (this.mark.CompareTo(obj.mark) > 0) return 1;
      else if (this.mark.CompareTo(obj.mark) < 0) return -1;
      // 2 поле
      else if (this.model.CompareTo(obj.model) > 0) return 1;
      else if (this.model.CompareTo(obj.model) < 0) return -1;
      // 3 поле
      else if (this.CPU_clockRate > obj.CPU_clockRate) return 1;
      else if (this.CPU_clockRate < obj.CPU_clockRate) return -1;
      // 4 поле
      else if (this.GPU_clockRate > obj.GPU_clockRate) return 1;
      else if (this.GPU_clockRate < obj.GPU_clockRate) return -1;
      // 5 поле
      else if (this.RAM_size > obj.RAM_size) return 1;
      else if (this.RAM_size < obj.RAM_size) return -1;
      // 6 поле
      else if (this.GPU_RAM_size > obj.GPU_RAM_size) return 1;
      else if (this.GPU_RAM_size < obj.GPU_RAM_size) return -1;

      return 0;
    }
      // Свойтва объектов
    private string
      mark,
      model;
    private float
      CPU_clockRate,
      GPU_clockRate;
    private int
      RAM_size,
      GPU_RAM_size;

      // Конструктор с параметрами
    public Laptop(
      string mark,
      string model,
      float CPU_clockRate,
      float GPU_clockRate,
      int RAM_size,
      int GPU_RAM_size
    ) {
      this.mark = mark;
      this.model = model;
      this.CPU_clockRate = CPU_clockRate;
      this.GPU_clockRate = GPU_clockRate;
      this.RAM_size = RAM_size;
      this.GPU_RAM_size = GPU_RAM_size;
    }

      // Конструктор без параметров
    public Laptop() {
      this.mark = "Empty";
      this.model = "Empty";
      this.CPU_clockRate = 0;
      this.GPU_clockRate = 0;
      this.RAM_size = 0;
      this.GPU_RAM_size = 0;
    }

    ~Laptop() {}

    public string GetMark() {
      return mark;
    }

    public string GetModel() {
      return model;
    }

    public float getCPUcr() {
      return CPU_clockRate;
    }

    public float GetGPUcr() {
      return GPU_clockRate;
    }

    public int GetRAMs() {
      return RAM_size;
    }

    public int GetGPUrs() {
      return GPU_RAM_size;
    }

    public string SetMark(string value) {
      mark = value;
      return value;
    }

    public string SetModel(string value) {
      model = value;
      return value;
    }

    public float SetCPUcr(float value) {
      CPU_clockRate = value;
      return value;
    }

    public float SetGPUcr(float value) {
      GPU_clockRate = value;
      return value;
    }

    public int SetRAMs(int value) {
      RAM_size = value;
      return value;
    }

    public int SetGPUrs(int value) {
      GPU_RAM_size = value;
      return value;
    }

  }
}
